### PR TITLE
Add fix for UnobservedTaskException when server Unavailable

### PR DESCRIPTION
**Pull request by rburns1293**
Checks the status code for Unavailable and observes the exception to avoid GC Finalizer from throwing it. 
Fixes [#24421](https://github.com/grpc/grpc/issues/24421)

@jtattermusch

### CODE REVIEW

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs [hunk 1]
```diff
@@ -648,7 +648,7 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
             if (status.StatusCode != StatusCode.OK)
             {
                 streamingResponseCallFinishedTcs.SetException(new RpcException(status, receivedStatus.Trailers));
-                if (status.StatusCode == StatusCode.Cancelled || origCancelRequested)
+                if (status.StatusCode == StatusCode.Cancelled || status.StatusCode == StatusCode.Unavailable || origCancelRequested)
                 {
                     // Make sure the exception set to the Task is observed,
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/AsyncCall.cs:653]: we could potentially reuse the  Task.ObserveException utility method from here: https://github.com/grpc/grpc-dotnet/pull/1660/files#diff-cf6772e9387e5b8b33dd040aed96eccded75c5937df6af33f242396871f802f2R32

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs [hunk 2]
```diff
@@ -648,13 +648,13 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
...
```
Comment (jtattermusch) [inline src/csharp/Grpc.Core/Internal/AsyncCall.cs:653]: For extra context, here's my understanding of why we're seeing the UnobservedException at all. 
I think the issue is with the way the streaming responses are processed.
The call.ResponseStream.MoveNext() reads the next response message and only once it  gets reaches a null response,
it observes the result of  `call.StreamingResponseCallFinishedTask` here:
https://github.com/grpc/grpc/blob/1cd6e69347cbf62a012477fe184ee6fa8f25d32c/src/csharp/Grpc.Core/Internal/ClientResponseStream.cs#L59

So if MoveNext() is not running (or if it doesn't exhaust the response stream) the exception coming from call.StreamingResponseCallFinishedTask will never be observed (which later shows up in finalizer as unobserved exception).

Note that the call.StreamingResponseCallFinishedTask "starts" as soon a the call is started (but there may or may not be a call to responseStream.MoveNext().

**Comment by jtattermusch** [comment]
Fix would need to be done against v1.46.x now that https://github.com/grpc/grpc/pull/29225 has been merged.

File: src/csharp/Grpc.IntegrationTesting/UnobservedTaskExceptionTest.cs [hunk 3]
```diff
@@ -96,5 +96,31 @@ public async Task NoUnobservedTaskExceptionForAbandonedStreamingResponse()
...
+                    }
+                }
+                catch (RpcException err)
+                {
+                    Console.WriteLine(err.Status.StatusCode);
+                }
+            await chan.ShutdownAsync();
+            GC.Collect();
+            GC.WaitForPendingFinalizers();
+        }
```
[the remark this answers is not in the document]
Comment (rburns1293) [reply src/csharp/Grpc.IntegrationTesting/UnobservedTaskExceptionTest.cs:128]: That would let the channel shutdown gracefully.

File: src/csharp/Grpc.Core/Internal/AsyncCall.cs [hunk 4]
```diff
@@ -648,13 +648,13 @@ private void HandleFinished(bool success, ClientSideStatus receivedStatus)
...
                     // otherwise this can trigger "Unobserved exception" when the response stream
                     // is not read until its end and the task created by the TCS is garbage collected.
                     // See https://github.com/grpc/grpc/issues/17458
-                    var _ = streamingResponseCallFinishedTcs.Task.Exception;
+                    streamingResponseCallFinishedTcs.Task.ObserveException();
```
[the remark this answers is not in the document]
Comment (rburns1293) [reply src/csharp/Grpc.Core/Internal/AsyncCall.cs]: SG!

File: src/csharp/Grpc.IntegrationTesting/UnobservedTaskExceptionTest.cs [hunk 5]
```diff
@@ -96,5 +96,31 @@ public async Task NoUnobservedTaskExceptionForAbandonedStreamingResponse()
 
             Assert.AreEqual(0, unobservedTaskExceptionCounter.Count);
         }
+        [Test]
+        public async Task NoUnobservedTaskExceptionForUnavailableServers()
+        {
+            TaskScheduler.UnobservedTaskException += (sender, args) => {
+                Assert.Fail("UnobservedTaskException hit.");
+            };
+            var chan = new Channel("localhost:100", ChannelCredentials.Insecure);
```
[the remark this answers is not in the document]
Comment (rburns1293) [reply src/csharp/Grpc.IntegrationTesting/UnobservedTaskExceptionTest.cs]: Ah yes it can.

**Comment by rburns1293** [comment]
Note: With the Grpc C# implementation being placed into maintenance mode, this PR will actually need to be remade for 1.46.x. We can complete the majority of the review here first before moving it.

**Pull request closed without merging**